psydac/feec/multipatch/examples_nc/h1_source_pbms_nc.py

```diff
@@ -0,0 +1,279 @@
+    print(' source_proj = {}'.format(source_proj))
+    print(' backend_language = {}'.format(backend_language))
+    print('---------------------------------------------------------------------------------------------------------')
+
+    print('building the multipatch domain...')
+    domain = build_multipatch_domain(domain_name=domain_name)
+    mappings = OrderedDict([(P.logical_domain, P.mapping) for P in domain.interior])
+    mappings_list = list(mappings.values())
+    ncells_h = {patch.name: [ncells[i], ncells[i]] for (i,patch) in enumerate(domain.interior)}
+    domain_h = discretize(domain, ncells=ncells_h)
+
+    print('building the symbolic and discrete deRham sequences...')
+    derham  = Derham(domain, ["H1", "Hcurl", "L2"])
+    derham_h = discretize(derham, domain_h, degree=degree)
+
+    # multi-patch (broken) spaces
+    V0h = derham_h.V0
+    V1h = derham_h.V1
+    V2h = derham_h.V2
+    print('dim(V0h) = {}'.format(V0h.nbasis))
+    print('dim(V1h) = {}'.format(V1h.nbasis))
+    print('dim(V2h) = {}'.format(V2h.nbasis))
+
+    print('broken differential operators...')
+    # broken (patch-wise) differential operators
+    bD0, bD1 = derham_h.broken_derivatives_as_operators
+    bD0_m = bD0.to_sparse_matrix()
+    # bD1_m = bD1.to_sparse_matrix()
+
+    print('building the discrete operators:')
+    print('commuting projection operators...')
+    nquads = [4*(d + 1) for d in degree]
+    P0, P1, P2 = derham_h.projectors(nquads=nquads)
+
+    I0 = IdLinearOperator(V0h)
+    I0_m = I0.to_sparse_matrix()
+
+    print('Hodge operators...')
+    # multi-patch (broken) linear operators / matrices
+    H0 = HodgeOperator(V0h, domain_h, backend_language=backend_language)
+    H1 = HodgeOperator(V1h, domain_h, backend_language=backend_language)
+
+    H0_m = H0.to_sparse_matrix()                # = mass matrix of V0
+    dH0_m  = H0.get_dual_Hodge_sparse_matrix()  # = inverse mass matrix of V0
+    H1_m = H1.to_sparse_matrix()                # = mass matrix of V1
+
+    print('conforming projection operators...')
+    # conforming Projections (should take into account the boundary conditions of the continuous deRham sequence)
+    cP0_m = construct_scalar_conforming_projection(V0h, hom_bc=[True,True])
+    # cP1_m = construct_vector_conforming_projection(V1h, hom_bc=[True, True])
+
+    if not os.path.exists(plot_dir):
+        os.makedirs(plot_dir)
+
+    def lift_u_bc(u_bc):
+        if u_bc is not None:
+            print('lifting the boundary condition in V0h...  [warning: Not Tested Yet!]')
+            # note: for simplicity we apply the full P1 on u_bc, but we only need to set the boundary dofs
+            u_bc = lambdify(domain.coordinates, u_bc)
+            u_bc_log = [pull_2d_h1(u_bc, m.get_callable_mapping()) for m in mappings_list]
+            # it's a bit weird to apply P1 on the list of (pulled back) logical fields -- why not just apply it on u_bc ?
+            uh_bc = P0(u_bc_log)
```

Comment:
Indeed there is an inconsistency here, which we should clarify.

On the one hand, the method `projectors` of the class `DiscreteDerham` (which works on single-patch domains) provides commuting projectors for the **physical** fields, as the pull-back is applied internally. See [here](https://github.com/pyccel/psydac/blob/cf4f4b6cc496457258aa444a92cb20d95af6c523/psydac/api/feec.py#L215-L226).

On the other hand, the method `projectors` of the class `DiscreteDerhamMultipatch` provides commuting projectors on the **logical** fields, as no pull-back operation is performed internally. See [here](https://github.com/pyccel/psydac/blob/ae90d77e4b7ee1f500661ee40ed56a8dedc370ee/psydac/feec/multipatch/api.py#L143-L153).

For reference, those methods use the internal classes `Projector_H1` and `Multipatch_Projector_H1`, which work on **logical** fields. These classes are in charge of performing (patch-wise) global interpolation in a parallel setting.

Reply:
see Issue #409